### PR TITLE
feat: Implemented automatic purging of old backups.

### PR DESCRIPTION
**Pull request by Hobart2967**
- Implemented Property "purge-after" which indicates after how many minutes a backup expires.
- Implemented a check on the expiry date on each backup performing.
- Implemented automatic purging of backups which have expired.

### CODE REVIEW

File: src/main/java/org/shanerx/backup/AutoBackup.java [hunk 1]
```diff
@@ -151,41 +160,77 @@ public void run() {
...
+                .filter(currentBackup -> currentBackup.getMode().equals(mode.getName()))
+                .collect(Collectors.toSet());
+
+        for (BackupFile backup : backupsForThisMode) {
+            LocalDateTime backupDate = backup.getBackupDate();
+
+            long backupTimeMilliSeconds = backupDate.toInstant(ZoneOffset.ofTotalSeconds(0)).toEpochMilli();
+            long millisecondsNow = LocalDateTime.now().toInstant(ZoneOffset.ofTotalSeconds(0)).toEpochMilli();
+            long purgeTime = backupTimeMilliSeconds + mode.getPurgeAfter() * 1000 * 60;
+            if (millisecondsNow < purgeTime) {
```
Comment (SparklingComet) [inline src/main/java/org/shanerx/backup/AutoBackup.java:218]: If purgeTime is 0 (disabled), the backup will be deleted.

File: src/main/java/org/shanerx/backup/BackupMode.java [hunk 2]
```diff
@@ -28,13 +28,15 @@ public class BackupMode {
...
+    private int purgeAfter;
 
-    public BackupMode(String name, File dir, boolean allowManual, int schedule, int compressionLevel, boolean recursive) {
+    public BackupMode(String name, File dir, boolean allowManual, int schedule, int compressionLevel, boolean recursive, int purgeAfter) {
         this.name = name;
         this.dir = dir;
         this.allowManual = allowManual;
         this.schedule = Math.max(schedule, 0); // if <= 0 set to 0, otherwise to predefined value
         this.recursive = recursive;
+        this.purgeAfter = purgeAfter;
```
Comment (SparklingComet) [inline src/main/java/org/shanerx/backup/BackupMode.java:40]: Might to the same thing here as for the `schedule` field.

File: src/main/resources/config.yml [hunk 3]
```diff
@@ -97,48 +97,55 @@ backup-modes:
     dir: '/'
     allow-manual: true
     schedule: 0
+    purge-after: 0
```
Comment (SparklingComet) [inline src/main/resources/config.yml:103]: Would be good to add a note to the above comment, saying

- The time unit of the `purge-after` time interval is _minutes_
- Set to `0` or negative to disable (see my other comments)

**Review by SparklingComet — CHANGES_REQUESTED** [review]
Hello there,
thanks a lot for your fancy commit :P -- I'll be testing it out asap!
Might I suggest the following (trivial) changes? Thanks ~

File: src/main/java/org/shanerx/backup/AutoBackup.java (hunk 1 above)
Comment (Hobart2967) [reply src/main/java/org/shanerx/backup/AutoBackup.java:218]: Hm, couldve sworn that this last fix was already in. Will do! :D

Comment (Hobart2967) [reply src/main/java/org/shanerx/backup/AutoBackup.java:218]: Nope, it's not! Have a look at line 203, there, I pre-check for auto-purging. if its smaller than or equal to zero, all autopurging for that mode is skipped :) So, no need to change and all fine :)

Comment (SparklingComet) [reply src/main/java/org/shanerx/backup/AutoBackup.java:218]: Ah! Missed that, thanks

File: src/main/java/org/shanerx/backup/BackupMode.java (hunk 2 above)
Comment (Hobart2967) [reply src/main/java/org/shanerx/backup/BackupMode.java:40]: Done!

File: src/main/resources/config.yml (hunk 3 above)
Comment (Hobart2967) [reply src/main/resources/config.yml:103]: Done!

**Comment by Hobart2967** [comment]
Done with the change requests :) Have fun re-reviewing and happy easter! 🐰 🐔 🥚 

PS: I also fixed a problem with easily upgrading the plugin: Resolving the purge-after is now a "try-resolve" with a default to 0. Not all users will read the changelog and receive a NullPointerException, because "map.get" will throw (BEcause of the still missing property in their config)

**Review by SparklingComet — APPROVED** [review]
Alright looks good to me! I'll give it a quick test spin and merge :D

Thanks a lot for the changes and Happy Easter to you too!

**Comment by Hobart2967** [comment]
You're welcome! Have fun testing, looking forward to see it in the next releases :) What I recognized btw: Sometimes, the scheduled backups do not run, although min-players is set to 0. Do you know why that is?

**Comment by SparklingComet** [comment]
> Sometimes, the scheduled backups do not run, although min-players is set to 0. Do you know why that is?

Hmm interesting, I was not aware of any issues... Are you sure this is not because the config was reloaded? When that happens, I believe that the timer resets.

**Comment by Hobart2967** [comment]
Nope, had it also before my changes and I rarely reload my config, because many plugins have issues with doing so. In the end, I am in favor of clean restarts anyways :D

**Comment by SparklingComet** [comment]
@Hobart2967 

Sorry for the delayed reply.
I have given it a test run and it works quite fine (with the exception of an NPE that I have fixed).

As far as the non-occurring backups are concerned, I have been able to witness that and it is indeed very strange.
I'll try to fix it before the next release.

**Comment by SparklingComet** [comment]
Alright merged, thanks again for the PR! :D

**Comment by Hobart2967** [comment]
That'll be great, thanks for accepting! 👍✌️

**Pull request merged**